### DISCUSSION TITLE
How to listen multiple FutureProvider?

### DISCUSSION THREAD

**Discussion by namnh-0652** [rrousselGit/riverpod#2952] [Q&A]
Hi, I am new in `Riverpod`.
How to handle multiple loading (or `FutureProvider`)?
For Ex: I have 3 api calls in home screen, How can I keep the loading until all APIs call are completed.

```dart
@riverpod
Future<Movie> movieDetail(MovieDetailRef ref, int movieId) async {
  return ref.watch(getMovieDetailUseCaseProvider(movieId));
}

@riverpod
Future<List<Movie>> similarMovies(SimilarMoviesRef ref, int movieId) async {
  return ref.watch(getSimilarUseCaseProvider(movieId));
}

@riverpod
Future<Casts> movieCasts(MovieCastsRef ref, int movieId) async {
  return ref.watch(getCastsUseCaseProvider(movieId));
}
```
I try to what these above 3 providers like
```dart
@riverpod
Future<bool> screenLoading(ScreenLoadingRef ref, int movieId) async {
  return ref.watch(movieDetailProvider(movieId)).isLoading ||
    ref.watch(similarMoviesProvider(movieId)).isLoading ||
    ref.watch(movieCastsProvider(movieId)).isLoading;
}
```
However, Consumer is called multiple times.
```dart
Consumer(builder: (context, ref, value) {
  final isLoading = ref.watch(screenLoadingProvider(widget.movieId)).value;
  print('build loading $isLoading'); // the value is same but it still rebuilt
  return isLoading == true ? const Center(child: CircularProgressIndicator()) : const SizedBox();
})
```

Any ideas to handle the case like this?

**Comment by eyNmXGfPTS23** [comment]
My first idea would be to create a third FutureProvider which contain the 3 others.

```
@riverpod
Future<Movie> movieDetail(MovieDetailRef ref, int movieId) async {
  return ref.watch(getMovieDetailUseCaseProvider(movieId));
}

@riverpod
Future<List<Movie>> similarMovies(SimilarMoviesRef ref, int movieId) async {
  return ref.watch(getSimilarUseCaseProvider(movieId));
}

@riverpod
Future<Casts> movieCasts(MovieCastsRef ref, int movieId) async {
  return ref.watch(getCastsUseCaseProvider(movieId));
}

@riverpod
Future<(Movie, List<Movie>, Casts)> providerName(providerNameRef ref, int movieId) async {
  final value1;
  final value2;
  final value3;
  try {
    value1 = ref.watch(movieDetailProvider(movieId).future);
  } catch (error) {
    // Handle the error: rethrow or do something else
  }
  
  // Do that for the 2 other and return it like this.
  return (value1, value2, value3);
}
```

Or you can use [AsyncNotifier ](https://docs-v2.riverpod.dev/docs/providers/notifier_provider) which is a bit more complicated but I think it suit your need.

**Reply by namnh-0652** [reply]
Oh, I found that when I opened multiple movie detail screens with `context.push(MovieDetailPage.routeLocation, extra: movieId)` (go_router)

Ex: open movie with id 1,2,3,4 (4 is now on top)

Then, the loading listen from 4 is correct transition from `true->false`, however, other movie detail page 1,2,3 still listen with `false` value. 
How to prevent the paused screens stop observing the `screenLoadingProvider`?

**Reply by eyNmXGfPTS23** [reply]
I am not sure I understand your problem. If you are passing the movieId as parameter, riverpod should detect and create an provider for each page. So, the loading will depends on the future of each provider of the movieId is unique. Can you provide a bit of your code as example.

**Reply by namnh-0652** [reply]
I means "from movie detail, I can open another movie detail (similar movies)", then the behind (onPaused) detail page is still observes the provider and call rebuild.

```dart
Consumer(builder: (context, ref, value) {
  final isLoading = ref.watch(screenLoadingProvider(widget.movieId)).value;
  print('build loading $isLoading -- ${widget.movieId}');
  return isLoading == true ? const Center(child: CircularProgressIndicator()) : const SizedBox();
})

The printed output:
build loading true -- 4
build loading false -- 3
build loading false -- 2
build loading false -- 1
build loading true -- 4
build loading false -- 4
```

However, I found it can be go_router known bug https://github.com/flutter/flutter/issues/132049#issuecomment-1673962209

**Reply by eyNmXGfPTS23** [reply]
Well I am sorry but I do not use riverpod the same way you are using it. If it can help you, I am using like this when I need to show a loader when entering the page.
```
class TestWidget extends ConsumerWidget {
  const TestWidget({super.key});

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    final AsyncValue test = ref.watch(provider);

    test.when(data: (_) => Container(),
        error: (_, __) => Container(),
        loading: CircularProgressIndicator.new);
  }
}
```

So I don't know if the .when is available for you but I recommended using it since it handle the logic of loading, error and success. Maybe it will solve the go_router problem, I don't know.

**Reply by namnh-0652** [reply]
Thanks for your support, so, incase multiple api calls, you will combine them like your idea above?
```dart
@riverpod
Future<(Movie, List<Movie>, Casts)> providerName(providerNameRef ref, int movieId) async {
  final value1;
  final value2;
  final value3;
  try {
    value1 = ref.watch(movieDetailProvider(movieId).future);
  } catch (error) {
    // Handle the error: rethrow or do something else
  }
  
  // Do that for the 2 other and return it like this.
  return (value1, value2, value3);
}
```

**Reply by eyNmXGfPTS23** [reply]
Yes probably
